source/common/http/conn_manager_utility.cc

```diff
@@ -148,12 +148,24 @@ bool ConnectionManagerUtility::shouldTraceRequest(
   case Http::TracingType::All:
     return true;
   case Http::TracingType::UpstreamFailure:
-    return request_info.failureReason() != Http::AccessLog::FailureReason::None;
+    return isTraceableFailure(request_info);
   }
 
   // Compiler enforces switch above to cover all the cases and it's impossible to be here,
   // but compiler complains on missing return statement, this is to make compiler happy.
   NOT_IMPLEMENTED;
 }
 
+bool ConnectionManagerUtility::isTraceableFailure(
```

Comment:
between ConnectionManagerUtility and ResponseFlagUtils i think second one is preferable, will move.